### ISSUE TITLE
Datatable+selectManyCheckbox: Malformed XML response on "update"

### ISSUE THREAD

**Issue by Exandra** [primefaces/primefaces#4515]
## 1) Environment

- PrimeFaces version: 7.0.RC2
- JSF: **Mojarra 2.3.5** (!)
- Application server + version: 
  - Wildfly 14.0.1.Final 
  - primefaces-test (jetty)
- Affected browsers: All

## 2) Expected behavior

If you click on the commandButton in the sample XHTML while running Mojarra 2.3.5, two forms should be updated.

The commandButton-update should generate a well-formed XML response, like in Mojarra 2.2, as follows:

```
<?xml version='1.0' encoding='utf-8'?>
<partial-response id="j_id1">
	<changes>
		<update id="form1">
			<![CDATA[
<form id="form1" name="form1" method="post" action="/primefaces-test/test.xhtml" enctype="application/x-www-form-urlencoded">
<input type="hidden" name="form1" value="form1" />
58:54<div id="form1:j_idt6" class="ui-datatable ui-widget"><div id="form1:j_idt6_paginator_top" class="ui-paginator ui-paginator-top ui-widget-header ui-corner-top" role="navigation" aria-label="Pagination"><a href="#" class="ui-paginator-first ui-state-default ui-corner-all ui-state-disabled" aria-label="First Page" tabindex="-1"><span class="ui-icon ui-icon-seek-first">F</span></a><a href="#" class="ui-paginator-prev ui-state-default ui-corner-all ui-state-disabled" aria-label="Previous Page" tabindex="-1"><span class="ui-icon ui-icon-seek-prev">P</span></a><span class="ui-paginator-pages"></span><a href="#" class="ui-paginator-next ui-state-default ui-corner-all ui-state-disabled" aria-label="Next Page" tabindex="-1"><span class="ui-icon ui-icon-seek-next">N</span></a><a href="#" class="ui-paginator-last ui-state-default ui-corner-all ui-state-disabled" aria-label="Last Page" tabindex="-1"><span class="ui-icon ui-icon-seek-end">E</span></a><label id="form1:j_idt6:j_id21_rppLabel" for="form1:j_idt6:j_id21" class="ui-paginator-rpp-label ui-helper-hidden">Rows Per Page</label><select id="form1:j_idt6:j_id21" name="form1:j_idt6_rppDD" aria-labelledby="form1:j_idt6:j_id21_rppLabel" class="ui-paginator-rpp-options ui-widget ui-state-default ui-corner-left" value="5" autocomplete="off"><option value="5" selected="selected">5</option><option value="10">10</option><option value="15">15</option><option value="30">30</option><option value="50">50</option></select></div><div class="ui-datatable-tablewrapper"><table role="grid"><thead id="form1:j_idt6_head"><tr role="row"></tr></thead><tbody id="form1:j_idt6_data" class="ui-datatable-data ui-widget-content"><tr class="ui-widget-content ui-datatable-empty-message"><td colspan="0">No records found.</td></tr></tbody></table></div><div id="form1:j_idt6_paginator_bottom" class="ui-paginator ui-paginator-bottom ui-widget-header ui-corner-bottom" role="navigation" aria-label="Pagination"><a href="#" class="ui-paginator-first ui-state-default ui-corner-all ui-state-disabled" aria-label="First Page" tabindex="-1"><span class="ui-icon ui-icon-seek-first">F</span></a><a href="#" class="ui-paginator-prev ui-state-default ui-corner-all ui-state-disabled" aria-label="Previous Page" tabindex="-1"><span class="ui-icon ui-icon-seek-prev">P</span></a><span class="ui-paginator-pages"></span><a href="#" class="ui-paginator-next ui-state-default ui-corner-all ui-state-disabled" aria-label="Next Page" tabindex="-1"><span class="ui-icon ui-icon-seek-next">N</span></a><a href="#" class="ui-paginator-last ui-state-default ui-corner-all ui-state-disabled" aria-label="Last Page" tabindex="-1"><span class="ui-icon ui-icon-seek-end">E</span></a><label id="form1:j_idt6:j_id22_rppLabel" for="form1:j_idt6:j_id22" class="ui-paginator-rpp-label ui-helper-hidden">Rows Per Page</label><select id="form1:j_idt6:j_id22" name="form1:j_idt6_rppDD" aria-labelledby="form1:j_idt6:j_id22_rppLabel" class="ui-paginator-rpp-options ui-widget ui-state-default ui-corner-left" value="5" autocomplete="off"><option value="5" selected="selected">5</option><option value="10">10</option><option value="15">15</option><option value="30">30</option><option value="50">50</option></select></div></div><script id="form1:j_idt6_s" type="text/javascript">PrimeFaces.cw("DataTable","widget_form1_j_idt6",{id:"form1:j_idt6",paginator:{id:['form1:j_idt6_paginator_top','form1:j_idt6_paginator_bottom'],rows:5,rowCount:0,page:0,currentPageTemplate:'({currentPage} of {totalPages})'},disableContextMenuIfEmpty:false});</script>
</form>]]>
		</update>
		<update id="form2">
			<![CDATA[
<form id="form2" name="form2" method="post" action="/primefaces-test/test.xhtml" enctype="application/x-www-form-urlencoded">
<input type="hidden" name="form2" value="form2" />
<button id="form2:j_idt7" name="form2:j_idt7" class="ui-button ui-widget ui-state-default ui-corner-all ui-button-text-only" onclick="PrimeFaces.ab({s:&quot;form2:j_idt7&quot;,f:&quot;form2&quot;,u:&quot;form1 form2&quot;});return false;" type="submit"><span class="ui-button-text ui-c">Execute Me!</span></button><script id="form2:j_idt7_s" type="text/javascript">PrimeFaces.cw("CommandButton","widget_form2_j_idt7",{id:"form2:j_idt7"});</script><table id="form2:j_idt8" role="presentation" class="ui-selectmanycheckbox ui-widget"><tr><td><div class="ui-chkbox ui-widget"><div class="ui-helper-hidden-accessible"><input id="form2:j_idt8:0" name="form2:j_idt8" type="checkbox" value="" checked="checked" /></div><div class="ui-chkbox-box ui-widget ui-corner-all ui-state-default ui-state-active ui-state-error"><span class="ui-chkbox-icon ui-icon ui-icon-check ui-c"></span></div></div><label for="form2:j_idt8:0">First</label></td><td><div class="ui-chkbox ui-widget"><div class="ui-helper-hidden-accessible"><input id="form2:j_idt8:1" name="form2:j_idt8" type="checkbox" value="" checked="checked" /></div><div class="ui-chkbox-box ui-widget ui-corner-all ui-state-default ui-state-active ui-state-error"><span class="ui-chkbox-icon ui-icon ui-icon-check ui-c"></span></div></div><label for="form2:j_idt8:1">Second</label></td></tr><tr><td><div class="ui-chkbox ui-widget"><div class="ui-helper-hidden-accessible"><input id="form2:j_idt8:2" name="form2:j_idt8" type="checkbox" value="" checked="checked" /></div><div class="ui-chkbox-box ui-widget ui-corner-all ui-state-default ui-state-active ui-state-error"><span class="ui-chkbox-icon ui-icon ui-icon-check ui-c"></span></div></div><label for="form2:j_idt8:2">Third</label></td></table><script id="form2:j_idt8_s" type="text/javascript">PrimeFaces.cw("SelectManyCheckbox","widget_form2_j_idt8",{id:"form2:j_idt8"});</script>
</form>]]>
		</update>
		<update id="j_id1:javax.faces.ViewState:0">
			<![CDATA[932654329926138516:5005521518837043206]]>
		</update>
		<extension ln="primefaces" type="args">{&#34;validationFailed&#34;:true}</extension>
	</changes>
</partial-response>
```

## 3) Actual behavior

The update response contains malformed XML, preventing the two forms to update themselves. Firefox throws an error on the console. Chrome just ignores the commandButton click.

```
<?xml version='1.0' encoding='utf-8'?>
<partial-response>
	<changes>
		<update id="form1">
			<![CDATA[
<form id="form1" name="form1" method="post" action="/primefaces-test/test.xhtml" enctype="application/x-www-form-urlencoded">
<input type="hidden" name="form1" value="form1" />
55:07<div id="form1:j_idt6" class="ui-datatable ui-widget"><div id="form1:j_idt6_paginator_top" class="ui-paginator ui-paginator-top ui-widget-header ui-corner-top" role="navigation" aria-label="Pagination"><a href="#" class="ui-paginator-first ui-state-default ui-corner-all ui-state-disabled" aria-label="First Page" tabindex="-1"><span class="ui-icon ui-icon-seek-first">F</span></a><a href="#" class="ui-paginator-prev ui-state-default ui-corner-all ui-state-disabled" aria-label="Previous Page" tabindex="-1"><span class="ui-icon ui-icon-seek-prev">P</span></a><span class="ui-paginator-pages"></span><a href="#" class="ui-paginator-next ui-state-default ui-corner-all ui-state-disabled" aria-label="Next Page" tabindex="-1"><span class="ui-icon ui-icon-seek-next">N</span></a><a href="#" class="ui-paginator-last ui-state-default ui-corner-all ui-state-disabled" aria-label="Last Page" tabindex="-1"><span class="ui-icon ui-icon-seek-end">E</span></a><label id="form1:j_idt6:j_id28_rppLabel" for="form1:j_idt6:j_id28" class="ui-paginator-rpp-label ui-helper-hidden">Rows Per Page</label><select id="form1:j_idt6:j_id28" name="form1:j_idt6_rppDD" aria-labelledby="form1:j_idt6:j_id28_rppLabel" class="ui-paginator-rpp-options ui-widget ui-state-default ui-corner-left" value="5" autocomplete="off"><option value="5" selected="selected">5</option><option value="10">10</option><option value="15">15</option><option value="30">30</option><option value="50">50</option></select></div><div class="ui-datatable-tablewrapper"><table role="grid"><thead id="form1:j_idt6_head"><tr role="row"></tr></thead><tbody id="form1:j_idt6_data" class="ui-datatable-data ui-widget-content"><tr class="ui-widget-content ui-datatable-empty-message"><td colspan="0">No records found.</td></tr></tbody></table></div><div id="form1:j_idt6_paginator_bottom" class="ui-paginator ui-paginator-bottom ui-widget-header ui-corner-bottom" role="navigation" aria-label="Pagination"><a href="#" class="ui-paginator-first ui-state-default ui-corner-all ui-state-disabled" aria-label="First Page" tabindex="-1"><span class="ui-icon ui-icon-seek-first">F</span></a><a href="#" class="ui-paginator-prev ui-state-default ui-corner-all ui-state-disabled" aria-label="Previous Page" tabindex="-1"><span class="ui-icon ui-icon-seek-prev">P</span></a><span class="ui-paginator-pages"></span><a href="#" class="ui-paginator-next ui-state-default ui-corner-all ui-state-disabled" aria-label="Next Page" tabindex="-1"><span class="ui-icon ui-icon-seek-next">N</span></a><a href="#" class="ui-paginator-last ui-state-default ui-corner-all ui-state-disabled" aria-label="Last Page" tabindex="-1"><span class="ui-icon ui-icon-seek-end">E</span></a><label id="form1:j_idt6:j_id29_rppLabel" for="form1:j_idt6:j_id29" class="ui-paginator-rpp-label ui-helper-hidden">Rows Per Page</label><select id="form1:j_idt6:j_id29" name="form1:j_idt6_rppDD" aria-labelledby="form1:j_idt6:j_id29_rppLabel" class="ui-paginator-rpp-options ui-widget ui-state-default ui-corner-left" value="5" autocomplete="off"><option value="5" selected="selected">5</option><option value="10">10</option><option value="15">15</option><option value="30">30</option><option value="50">50</option></select></div></div><script id="form1:j_idt6_s" type="text/javascript">PrimeFaces.cw("DataTable","widget_form1_j_idt6",{id:"form1:j_idt6",paginator:{id:['form1:j_idt6_paginator_top','form1:j_idt6_paginator_bottom'],rows:5,rowCount:0,page:0,currentPageTemplate:'({currentPage} of {totalPages})'},disableContextMenuIfEmpty:false});</script>
</form>]]>
		</update>
		<update id="form2">
			<![CDATA[
<form id="form2" name="form2" method="post" action="/primefaces-test/test.xhtml" enctype="application/x-www-form-urlencoded">
<input type="hidden" name="form2" value="form2" />
<button id="form2:j_idt7" name="form2:j_idt7" class="ui-button ui-widget ui-state-default ui-corner-all ui-button-text-only" onclick="PrimeFaces.ab({s:&quot;form2:j_idt7&quot;,f:&quot;form2&quot;,u:&quot;form1 form2&quot;});return false;" type="submit"><span class="ui-button-text ui-c">Execute Me!</span></button><script id="form2:j_idt7_s" type="text/javascript">PrimeFaces.cw("CommandButton","widget_form2_j_idt7",{id:"form2:j_idt7"});</script><table id="form2:j_idt8" role="presentation" class="ui-selectmanycheckbox ui-widget"><tr><td><div class="ui-chkbox ui-widget"><div class="ui-helper-hidden-accessible"><input id="form2:j_idt8:0" name="form2:j_idt8" type="checkbox" value="" checked="checked" /></div><div class="ui-chkbox-box ui-widget ui-corner-all ui-state-default ui-state-active ui-state-error"><span class="ui-chkbox-icon ui-icon ui-icon-check ui-c"></span></div></div><label for="form2:j_idt8:0">First</label></td><td><div class="ui-chkbox ui-widget"><div class="ui-helper-hidden-accessible"><input id="form2:j_idt8:1" name="form2:j_idt8" type="checkbox" value="" checked="checked" /></div><div class="ui-chkbox-box ui-widget ui-corner-all ui-state-default ui-state-active ui-state-error"><span class="ui-chkbox-icon ui-icon ui-icon-check ui-c"></span></div></div><label for="form2:j_idt8:1">Second</label></td></tr><tr><td><div class="ui-chkbox ui-widget"><div class="ui-helper-hidden-accessible"><input id="form2:j_idt8:2" name="form2:j_idt8" type="checkbox" value="" checked="checked" /></div><div class="ui-chkbox-box ui-widget ui-corner-all ui-state-default ui-state-active ui-state-error"><span class="ui-chkbox-icon ui-icon ui-icon-check ui-c"></span></div></div><label for="form2:j_idt8:2">Third</label></td></tr><script id="form2:j_idt8_s" type="text/javascript">PrimeFaces.cw("SelectManyCheckbox","widget_form2_j_idt8",{id:"form2:j_idt8"});</script>
</table>]]>
		</form>
		<update id="j_id1:javax.faces.ViewState:0">
			<![CDATA[1819805144490008794:7620409190201286769]]>
		</update>
		<extension ln="primefaces" type="args">{&#34;validationFailed&#34;:true}</extension>
	</update>
</partial-response>
```

Last `</update>` should be before `<update id="j_id1:javax.faces.ViewState:0">`. `</changes>` is missing.

## 4) Steps to reproduce

1. `git clone https://github.com/Exandra/primefaces-test`
2. `cd primefaces-test`
3. `mvn clean -P mojarra23 jetty:run-exploded`
4. Open http://localhost:8080/primefaces-test/ in your browser
5. Click on the commandButton 'Execute Me!'. The counter at the top of the page is not updated.
6. Compare your result with `mvn clean -P mojarra22 jetty:run-exploded`, the counter at the top of the page updates.

## 5) Sample XHTML

```
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml"
      xmlns:ui="http://java.sun.com/jsf/facelets"
      xmlns:f="http://java.sun.com/jsf/core"
      xmlns:p="http://primefaces.org/ui"
      xmlns:h="http://java.sun.com/jsf/html">

    <h:head>
        <title>PrimeFaces Test</title>
    </h:head>
    <h:body>

    <h:form id="form1">
        <h:outputText value="#{session.lastAccessedTime}">
            <f:convertDateTime pattern="mm:ss" type="date" />
        </h:outputText>

        <p:dataTable paginator="true" rows="5"
            rowsPerPageTemplate="5,10,15,30,50">
        </p:dataTable>
    </h:form>

    <h:form id="form2">
        <p:commandButton value="Execute Me!" update=":form1,@form" />

        <p:selectManyCheckbox layout="grid" columns="2">
            <f:selectItem itemLabel="First" />
            <f:selectItem itemLabel="Second" />
            <f:selectItem itemLabel="Third" />
        </p:selectManyCheckbox>
    </h:form>

    </h:body>
</html>
```

I do not know why this special combination is problematic. Try to remove `rowsPerPageTemplate="…"` or `layout="grid"` or update only one of both forms and everything works fine in Mojarra 2.3.

**Comment by tandraschko** [comment]
@melloware can you check this pls? Could be related to the other issue.

**Comment by melloware** [comment]
Let me look.  This would be very bizarre.

**Comment by melloware** [comment]
OK I have confirmed this is broken in 6.2 and 7.0.RC2.  Then I verified its fixed in 7.0-master.

@Exandra try this sample project: 
[pf-4515.zip](https://github.com/primefaces/primefaces/files/2842802/pf-4515.zip)

Just run "mvn clean jetty:run-exploded" and navigate to http://localhost:8080/primefaces-test/test.xhtml

I believe it was fixed with this bug: https://github.com/primefaces/primefaces/issues/4474

**Comment by tandraschko** [comment]
Thanks!
